doc/R-vignette/igraph_extended_tutorial_R_vignette_english.Rmd

```diff
@@ -0,0 +1,621 @@
+
+## Setting and retrieving attributes
+
+As mentioned above, in igraph each vertex and each edge have a numeric id from `1` upwards. Deleting vertices or edges can therefore cause reassignments of vertex and/or edge IDs. In addition to IDs, vertex and edges can have *attributes* such as a name, coordinates for plotting, metadata, and weights. The graph itself can have such attributes too (e.g. a name, which will show in `print` or `summary`). In a sense, every `Graph`, vertex and edge can be used as a R variable to store and retrieve these attributes.
+
+To demonstrate the use of attributes, let us create a simple social network:
+
+```{r}
+gi <- make_empty_graph()
+gi <- graph(edges = c(1,2, 1,3, 3,4, 4,5, 5,3, 3,6, 6,1, 7,4, 6,7), directed = FALSE)
+```
+
+Each vertex represents a person, so we want to store names, ages and genders:
+
+```{r echo = TRUE}
+V(gi)$name <- c("Alice", "Bob", "Claire", "Dennis", "Esther", "Frank", "George")
+V(gi)$age <- c(25, 31, 18, 47, 22, 23, 50) 
+V(gi)$gender <- c("f", "m", "f", "m", "f", "m", "m")
+E(gi)$is_formal <- c(FALSE, FALSE, TRUE, TRUE, TRUE, FALSE, TRUE, FALSE, FALSE)
+summary(gi)
+```
+
+`V` and `E` are the standard way to obtain a sequence of all vertices and edges, respectively. The value must be a list with the same length as the vertices (for `V`) or edges (for `E`). This assigns an attribute to *all* vertices/edges at once.
+
+To assign or modify an attribute for a single vertex/edge, you can use indexing:
+
+```{r echo = TRUE}
+E(gi)$is_formal
+```
+
+```{r}
+E(gi)$is_formal[1] <- c(TRUE)
+```
+
+```{r echo = TRUE}
+E(gi)$is_formal
+```
+
+In fact, a single vertex is represented via the class `Vertex`, and a single edge via `Edge`. Both of them plus `Graph` can all be keyed like a dictionary to set attributes, e.g. to add a date to the graph:
+
+```{r echo = TRUE}
+gi <- set_graph_attr(gi, "date", "2022-02-11")
+graph_attr(gi, "date")
+```
+
+To retrieve a dictionary of attributes, you can use `graph_attr`, `vertex_attr`, and `edge_attr`. To find the ID of a vertex you can use the function `match`:
+
+```{r echo = TRUE}
+match(c("George"),V(gi)$name)
+```
+
+To assign attributes to a subset of vertices or edges, you can use slicing:
+
+```{r echo = TRUE}
+V(gi)$name[1:3] <- c("Alejandra", "Bruno", "Carmina")
+V(gi)
+```
+
+To delete attributes:
+
+```{r echo = TRUE}
+gi <- delete_vertex_attr(gi,"gender")
+V(gi)$gender
+```
+
+```{=html}
+<!---
+Does this warning also apply to R?: Attributes can be arbitrary Python objects, but if you are saving graphs to a file, only string and numeric attributes will be kept. See the `pickle` module in the standard Python library if you are looking for a way to save other attribute types. You can either pickle your attributes individually, store them as strings and save them, or you can pickle the whole `Graph` if you know that you want to load the graph back into Python only.
```

Comment:
It depends on how you save the graph in R. Contrary to the Python interface, where the graph primarily lives on the C side and the Python object only holds a pointer to it, the graph is a first-class R object in the R interface; you can even inspect its internals by calling `unclass(gi)` on it. When a core C function is invoked on an igraph graph in R, the lists in the R graph object are cast into vector _views_ and they are forwarded to the C core in a zero-copy manner. The upside is that you can simply save an igraph graph in R with R's standard `dput()` method and retrieve it later with `dget()` without losing any information. You can also just save the R workspace and restore it later.

However, if you save the graph in R into a standard graph format like GraphML, you are limited to what the graph format can represent. There is no graph format in igraph that could save arbitrary R objects, so if the user wants to have all sorts of R attributes on the graph and preserve them, she should use R's `dput()` and `dget()` functions to save the graph in an R-specific format instead.